File: leetcode/200/main.js
class Queue {
  constructor() {
    this.q = [];
    this.head = 0;
  }
  push(x) {
    this.q.push(x);
  }
  pop() {
    return this.q[this.head++];
  }
  size() {
    return this.q.length - this.head;
  }
}

const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  const m = grid.length,
    n = grid[0].length;
  const isIn = (r, c) => 0 <= r && r < m && 0 <= c && c < n;
  const visited = Array.from({ length: m }, () => Array(n).fill(false));

  const findIsland = (sr, sc) => {
    const q = new Queue();
    q.push([sr, sc]);
    visited[sr][sc] = true;
    while (q.size()) {
      const [cr, cc] = q.pop();
      for (let i = 0; i < 4; i++) {
        const nr = cr + dx[i],
          nc = cc + dy[i];
        if (!isIn(nr, nc) || grid[nr][nc] !== "1" || visited[nr][nc]) continue;
        q.push([nr, nc]);
        visited[nr][nc] = true;
      }
    }
  };

  let answer = 0;
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] !== "1" || visited[i][j]) continue;
      answer++;
      findIsland(i, j);
    }
  }
  return answer;
};

File: leetcode/200/main.cpp
#include <queue>
#include <vector>
using namespace std;

using pii = pair<int, int>;

const int dx[4] = { -1, 0, 1, 0 };
const int dy[4] = { 0, 1, 0, -1 };

class Solution {
  int m, n;

  bool is_in(int r, int c) {
    return 0 <= r && r < m && 0 <= c && c < n;
  }

  void find_island(int r, int c, vector<vector<bool>> &is_visited, const vector<vector<char>> &grid) {
    queue<pii> q;
    q.push({ r, c }), is_visited[r][c] = true;

    while (!q.empty()) {
      auto [cr, cc] = q.front();
      q.pop();

      for (int i = 0; i < 4; i++) {
        int nr = cr + dx[i], nc = cc + dy[i];
        if (!is_in(nr, nc) || grid[nr][nc] != '1' || is_visited[nr][nc]) continue;
        q.push({ nr, nc }), is_visited[nr][nc] = true;
      }
    }
  }

public:
  int numIslands(vector<vector<char>> &grid) {
    this->m = grid.size(), this->n = grid[0].size();
    int answer = 0;
    vector<vector<bool>> is_visited(m, vector<bool>(n, false));

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (grid[i][j] != '1' || is_visited[i][j]) continue;

        answer += 1;
        find_island(i, j, is_visited, grid);
      }
    }

    return answer;
  }
};
